### PR TITLE
Added board.is_valid_board() function

### PR DESCRIPTION
**Pull request by RandyBrown1965**
board.is_valid_board() tests the board layout and ensures that it agrees with the board. dim and board.num_agents and that it the shape of the wires on it is valid (with each having exactly one head and one target).

### CODE REVIEW

File: ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+#from jax.numpy import asarray  Currently jaxlib is not supported on windows.  This will have to be sorted.
+#from env_viewer import RoutingViewer  Currently jaxlib is not supported on windows.  This will have to be sorted.
+
+@dataclass
+class Position:
+    # Class of 2D tuple of ints, indicating the size of an array or a 2D position or vector.
+    x: int
+    y: int
+
+class Board:
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py:21]: Please rename so that we dont have multiple classes called board within our repo

File: ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+        connectible_list.append((x_pos, y_pos))
+        # Recursively add the cells above, to the right, below, and to the left if they're valid and open
+        if (x_pos < self.dim.x-1) and (self.layout[x_pos+1, y_pos] == EMPTY) and (x_pos+1, y_pos) not in connectible_list:
+            self.add_connectible_cell(x_pos+1, y_pos, connectible_list)
+        if (y_pos < self.dim.y-1) and (self.layout[x_pos, y_pos+1] == EMPTY) and (x_pos, y_pos+1) not in connectible_list:
+            self.add_connectible_cell(x_pos, y_pos+1, connectible_list)
+        if (x_pos > 0) and (self.layout[x_pos-1, y_pos] == EMPTY) and (x_pos-1, y_pos) not in connectible_list:
+            self.add_connectible_cell(x_pos-1, y_pos, connectible_list)
+        if (y_pos > 0) and (self.layout[x_pos, y_pos-1] == EMPTY) and (x_pos, y_pos-1) not in connectible_list:
+            self.add_connectible_cell(x_pos, y_pos-1, connectible_list)
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py:268]: feels like (x_pos < self.dim.x-1) and (self.layout[x_pos+1, y_pos] == EMPTY) and (x_pos+1, y_pos) not in connectible_list could be wrapped in a function and you pass the arguments that you want to evaluate

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+from typing import List, Tuple
+#from jax.numpy import asarray # Currently jaxlib is not supported on windows.  This will have to be sorted.
+#from env_viewer import RoutingViewer # Currently jaxlib is not supported on windows.  This will have to be sorted.
+
+@dataclass#
+class Position:
+    # Class of 2D tuple of ints, indicating the size of an array or a 2D position or vector.
+    x: int
+    y: int
+
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:20]: position should only be in one file and then imported into each file - I think I put one in data model (I know you have import issues at the moment but this will need refactoring)

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+class Board:
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:21]: As before, please rename class

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    Heads are encoded starting from 4 in multiples of 3: 4, 7, 10, ...
+    Targets are encoded starting from 3 in multiples of 3: 3, 6, 9, ...
+    Wiring routes connecting the head/target pairs are encoded starting
+            at 2 in multiples of 3: 2, 5, 8, ...
+
+    Args:
+        rows, cols (int, int) : Dimensions of the board.
+
+    """
+    def __init__(self, rows: int, cols: int):
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:36]: should take num_agents in init

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py (hunk 4 above)
Comment (mwolinska) [reply ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:21]: you can also consider inheriting from a version and only changing the code that is different rather than reimplementing the whole thing - I did something similar for Ugo's boards if you want to see what I mean

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+    def add_connectible_cell(self, x_pos: int, y_pos: int, connectible_list: List) -> List:
+        """ Add the specified cell to the list, recursively call adjacent cells, and return list.
+
+        Args:
+            x_pos, y_pos (int, int) : cell to add to the list.
+            connectible_list (List[(int,int)...] : input list of connected cells.
+
+        Returns:
+            List[(int,int)...] : output list of connected cells.
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:255]:
```suggestion
            List[Tuple[int, int]] : output list of connected cells.
```

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        if (y_pos > 0) and (self.layout[x_pos, y_pos-1] == EMPTY) and (x_pos, y_pos-1) not in connectible_list:
+            self.add_connectible_cell(x_pos, y_pos-1, connectible_list)
+        return connectible_list
+
+    def is_connectible(self, x_head: int, y_head: int, x_target: int, y_target: int) -> bool:
+            """ Return a boolean indicating if the two cells are connectible on the board.
+
+        Args:
+            x_head, y_head (int, int) : cell at one end of the proposed wire.
+            x_target, y_target (int, int) : cell at the other end of the proposed wire.
```
Comment (mwolinska) [inline ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:276]: I think these should be in separate lines since they are individual arguments

**Review by mwolinska — CHANGES_REQUESTED** [review]
Before merging
1. rebase before merging 
2. rename boards (I think I had some proposals in the current version on dev
3. Add them to the interface

File: ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py (hunk 1 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py:21]: Done

File: ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py (hunk 2 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_0_0_rb.py:268]: Done

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py (hunk 3 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:20]: As discussed, there will only be one file.  Old version are in the obsolete sub-folder.

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py (hunk 5 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:36]: Done.
Let's please revisit this with the group.

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py (hunk 6 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:255]: Done

File: ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py (hunk 7 above)
Comment (RandyBrown1965) [reply ic_routing_board_generation/board_generator/board_generator_v3_1_0_rb.py:276]: As discussed, I have changed the comment to indicate that these two form a 2D position.
